cpp/cugraph_module/algorithms/graph_generator.cu

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2016-2022 Memgraph Ltd. [https://memgraph.com]
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "mg_cugraph_utility.hpp"
+
+namespace {
+using vertex_t = int64_t;
+using edge_t = int64_t;
+using weight_t = double;
+
+constexpr char const *kProcedureGenerate = "rmat";
+
+constexpr char const *kArgumentScale = "scale";
+constexpr char const *kArgumentNumEdges = "num_edges";
+constexpr char const *kArgumentParameterA = "a";
+constexpr char const *kArgumentParameterB = "b";
+constexpr char const *kArgumentParameterC = "c";
+constexpr char const *kArgumentSeed = "seed";
+constexpr char const *kArgumentClipAndFlip = "clip_and_flip";
+
+constexpr char const *kFieldMessage = "message";
+
+void InsertMessageRecord(mgp_result *result, mgp_memory *memory, const char *message) {
+  auto *record = mgp::result_new_record(result);
+
+  mg_utility::InsertStringValueResult(record, kFieldMessage, message, memory);
+}
+
+struct VertexDelete {
+  void operator()(mgp_vertex *v) {
+    if (v) mgp::vertex_destroy(v);
+  }
+};
+
+void GenerateRMAT(mgp_list *args, mgp_graph *graph, mgp_result *result, mgp_memory *memory) {
+  try {
+    auto scale = mgp::value_get_int(mgp::list_at(args, 0));
+    auto num_edges = mgp::value_get_int(mgp::list_at(args, 1));
+    auto parameter_a = mgp::value_get_double(mgp::list_at(args, 2));
+    auto parameter_b = mgp::value_get_double(mgp::list_at(args, 3));
+    auto parameter_c = mgp::value_get_double(mgp::list_at(args, 4));
+    auto seed = mgp::value_get_int(mgp::list_at(args, 5));
+    auto clip_and_flip = mgp::value_get_bool(mgp::list_at(args, 6));
+
+    raft::handle_t handle{};
+
+    auto num_vertices = 1 << scale;  // RMAT generator defines this
+    auto edges = mg_cugraph::GenerateCugraphRMAT(scale, num_edges, parameter_a, parameter_b, parameter_c, seed,
+                                                 clip_and_flip, handle);
+
+    std::vector<std::unique_ptr<mgp_vertex, VertexDelete>> vertices(num_vertices);
+    for (std::size_t i = 0; i < num_vertices; ++i) {
+      auto new_vertex = mgp::graph_create_vertex(graph, memory);
+
+      // Add labels as arguments
+      vertices[i] = std::unique_ptr<mgp_vertex, VertexDelete>(mgp::vertex_copy(new_vertex, memory));
```

Comment:
I don't know C++ the best, but wouldn't `std::move` here work?

Reply:
The problem is that when the local variable new_vertex gets destroyed, the vertices[i] will be destroyed too. We should talk to our C++ gurus once about this.

Reply:
But what if we don't destroy then the variable down below?